### PR TITLE
doc: #9 add apidocs

### PR DESCRIPTION
**Pull request by brunopacheco1**

### CODE REVIEW

**Review by nolliia — APPROVED** [review]

File: apidocs.yaml [hunk 1]
```diff
@@ -0,0 +1,826 @@
...
+          application/json:
+            schema:
+              $ref: "#/components/schemas/RemoveMember"
+      responses:
+        "204":
+          description: Successful Response (no content)
+      security:
+        - auth:
+          - write:applications
+  /api/v1/applications/{id}/manage-datasets:
```
Comment (admy7) [inline apidocs.yaml:344]: I am not sure to understand the action implemented by this endpoint, especially the meaning of "manage". 

Does that mean adding (linking) datasets to a certain application?

File: apidocs.yaml [hunk 2]
```diff
@@ -0,0 +1,826 @@
...
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: "#/components/schemas/ValidationWarnings"
+      security:
+        - auth:
+          - write:applications
+  /api/v1/applications/create-application:
```
Comment (inderps) [inline apidocs.yaml:137]: isnt just having POST `/api/v1/applications` more "restful" than POST `/api/v1/applications/create-application` for creation?

File: apidocs.yaml [hunk 3]
```diff
@@ -0,0 +1,826 @@
...
+          required: true
+          schema:
+            type: string
+      responses:
+        "204":
+          description: Successful Response (no content)
+      security:
+        - auth:
+          - write:applications
+  /api/v1/applications/{id}/copy-as-new:
```
Comment (inderps) [inline apidocs.yaml:181]: is it duplicate or something else?

File: apidocs.yaml [hunk 4]
```diff
@@ -0,0 +1,826 @@
...
+          required: true
+          schema:
+            type: string
+      responses:
+        "204":
+          description: Successful Response (no content)
+      security:
+        - auth:
+          - write:applications
+  /api/v1/applications/{id}/add-attachment:
```
Comment (inderps) [inline apidocs.yaml:219]: This one too: POST `/api/v1/applications/{id}/attachments` will be more "restful" no?

File: apidocs.yaml [hunk 5]
```diff
@@ -0,0 +1,826 @@
...
+            type: string
+      responses:
+        "200":
+          description: Successful Response
+          content:
+            application/pdf:
+              schema:
+                type: string
+                format: binary
+  /api/v1/applications/{id}/remove-attachment:
```
Comment (inderps) [inline apidocs.yaml:277]: another alternative: DELETE `/api/v1/applications/{id}`

File: apidocs.yaml [hunk 6]
```diff
@@ -0,0 +1,826 @@
...
+          application/json:
+            schema:
+              $ref: "#/components/schemas/DeleteAttachment"
+      responses:
+        "204":
+          description: Successful Response (no content)
+      security:
+        - auth:
+          - write:applications
+  /api/v1/applications/{id}/add-member:
```
Comment (inderps) [inline apidocs.yaml:301]: POST `/api/v1/applications/{id}/members`

Then you can also have GET `/api/v1/applications/{id}/members` to get all members for application no?

File: apidocs.yaml (hunk 2 above)
Comment (brunopacheco1) [reply apidocs.yaml:137]: It would make sense if we were posting an application, but this endpoint creates an application from a list of selected datasets.

What will do behind the seen is fetching from DAAM the configuration of an application (forms and their fields, licenses, duo, etc...), fill a object of type Application and assign it to the datasets.

File: apidocs.yaml (hunk 4 above)
Comment (brunopacheco1) [reply apidocs.yaml:219]: I will close this conversation to keep it in a single thread.

File: apidocs.yaml (hunk 3 above)
Comment (brunopacheco1) [reply apidocs.yaml:181]: Nope, this is an action or a command to copy an existing application as a new one. Do you think this endpoint name is not appropriate?

File: apidocs.yaml (hunk 1 above)
Comment (admy7) [reply apidocs.yaml]: For my own understanding, can you explain me why you did so instead of creating multiple endpoints with specific http verbs for adding, deleting and updating datasets? 
Like you did for /api/v1/applications/{id}/remove-member and /api/v1/applications/{id}/add-member

It is because we are dealing with a collection here? I must admit, that would become cumbersome if we had to do 
DELETE /api/v1/applications/{id}/{dataset_id}

**Review by admy7 — APPROVED** [review]

File: apidocs.yaml (hunk 6 above)
Comment (brunopacheco1) [reply apidocs.yaml:301]: I must discuss this face to face with the team, but the reasons are multiple.

The problem: A RESTfull API is a basic CRUD, but it is too rigid and aggregates a lot of different responsibilities into too few endpoints, which very often do not reflect business scenarios and use cases. It's almost like a one to one between HTTP call and database queries.

So I wanted to prepare the ground and introduce different principles: event-driven design, DDD, Clean Archictecture, CQRS, Data-Oriented Programming, SOLID...

Basically, I tried to break the API into functional endpoints / use cases, so you have in the endpoint path clear indication on what it is doing. We can have specific actions, Commands and Queries, and they will have again very specialised resquest and response payload, with the most meaningfull data for each action.

With that we reduce coupling and allows the API to evolve and introduce different technologies or features with minimum impact in the overal system.

Comment (brunopacheco1) [reply apidocs.yaml:301]: Many things you pointed are valid, but they do not cover business scenarios.

Like `POST /application/{id}/members` or `DELETE /application/{id}/members/{memberId}`.

The first endpoint does not reflect the idea that a member is not directly added to the application, but actually a notification should be send to the member and they have to accept the invitation, and only then it's added to the application.

The second does not reflect the idea that a reason must be provided for the member to be removed, and in a RESTfull API DELETE verb usually does not come with a body.

File: apidocs.yaml (hunk 1 above)
Comment (brunopacheco1) [reply apidocs.yaml]: I will close this conversation to keep this topic in a single thread.

https://github.com/GenomicDataInfrastructure/gdi-userportal-access-management-service/pull/14#discussion_r1484338971

File: apidocs.yaml (hunk 3 above)
Comment (inderps) [reply apidocs.yaml:181]: no its fine

File: apidocs.yaml (hunk 6 above)
Comment (inderps) [reply apidocs.yaml]: Indeed, for special cases like` invite-member` or delete member with body, deviating makes sense. 

However, for endpoints where it's applicable, shouldn't we adhere to the REST convention? 

I personally have preference for the REST convention because it simplifies naming; it's straightforward and intuitive, You don't have to think much. it just comes naturally

Nevertheless, I don't have a strong opinion on this matter. I'm open to using custom routes as well

**Review by admy7 — DISMISSED** [review]

File: apidocs.yaml [hunk 7]
```diff
@@ -0,0 +1,849 @@
...
+        "200":
+          description: Successful Response
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/RetrievedApplication"
+      security:
+        - auth:
+          - read:applications
+  /api/v1/applications/{id}/save-forms-and-duos:
```
Comment (inderps) [inline apidocs.yaml:88]: what is duo here?

**Review by inderps — APPROVED** [review]

File: apidocs.yaml (hunk 7 above)
Comment (brunopacheco1) [reply apidocs.yaml:88]: Thanks for spotting it. It's fixed. It's another part of the application that is saved at the same moment the forms are saved.

File: apidocs.yaml (hunk 6 above)
Comment (brunopacheco1) [reply apidocs.yaml]: Indeed, I will follow REST convention whenever is possible, for intance, all `GET` endpoints are RESTful. 

And I changed: 
- `POST /applications/{id}/add-attachement` to `POST /applications/{id}/attachments`.
- `POST /applications/{id}/add-event` to `POST /applications/{id}/events`.

All the other endpoints do not fit on RESTful convention because they have extra business steps.

The only special case for me is `/applications/{id}/save-forms-and-duos`. It could be simply `PUT /applications/{id}/forms` and `PUT /applications/{id}/duos`, but I don't know the impact on the integration with REMS. In REMS both forms and DUO codes are persisted together.

**Review by inderps — APPROVED** [review]

**Review by admy7 — APPROVED** [review]
